### PR TITLE
[solax] Add support for Solax X3MIC / G2 inverter and workmode channel

### PR DESCRIPTION
**Pull request by realthk**
Support for my Solax X3MIC / G2 inverter, and include workmode which is useful if the inverter cannot work continuously (for example our area very often has over-voltage problem)

### CODE REVIEW

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxBindingConstants.java
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxBindingConstants.java:64]: Any clue what temperature 1 and 2 means? Could be nice if we can somehow make it more explicit if we know it. If we don't know it, I guess it's OK. :) (same applies for the metadata files)

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxBindingConstants.java
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxBindingConstants.java:109]: Best would be to use an ENUM for these values. See the InverterType enum for example.

Could be even better to use these as I18N Keys constants so they get translated if they're used in the UI with a different locale, i.e. make the values all caps and translate them wherever you pass them to the UI/framework as states.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java:204]: Not sure if Units.ONE is the most proper one. Maybe a good question to the approving reviewer. For me it's OK.

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/model/InverterData.java
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/model/InverterData.java:100]: I just checked the inverter status values in the Cloud API document. From what I can see there are codes 100-113 and the first 6 are just like you parsed them, i.e. 100 is Wait, 101 is Checking. Maybe we could assume that the other codes are the same as here but without the first 1 at the beginning. Could be a nice subsequent feature request...

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/model/InverterData.java
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/model/InverterData.java:114]: Add a default state instead of returning a null value.
```suggestion
       default: return WORKMODE_UNKNOWN;
        }
```

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/model/InverterData.java
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/model/InverterData.java:101]: As suggested, maybe return an ENUM here instead of a string constant. Makes things more predictable.

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/channel_types.xml:43]: Isn't this workmode actually an inverter status? I guess both words could work but which one is more precise?

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:70]: Nice catch! Thanks! :) (applies for the others too)

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml
Comment (theater) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:117]: See my comment for the constant - same applies here (even more than on the constant layer as this is presented to the user)

**Review by theater — COMMENTED** [review]
Hi.
Things look pretty much OK to me. I added some comments.
What is missing is an update instructions for the added channels.
See how it's done in the local_connect_inverter_type_update.xml
I don't request changes as I'm not an official reviewer so I added only a comments :)

Cheers,
Konstantin

**Comment by realthk** [comment]
Thanks Konstantin!
Sorry for the late reply, I'm quite busy with my daily job now, and have to admit, OpenHab is not my main system, only trying it for curiosity. I have my Solax X3MicG2 inverter integrated into Home Assistant by transforming data fetched by REST with Jinja expressions in simple template sensors. 
Also, besides OpenHAB is new for me, I have no experience with Java at all, just duplicated an original class for other inverter, and modified where and how it seemed logical until it was working fine on my system. So this is the explanation why I used Units.ONE, and you're perfectly right, I also felt it fishy, but had no better idea for passing a unit-less state number. Same goes for enum and translation, I agree completely, just did not know how to do better. 
As for the two temperature sensors, I could not find any documentation, so yes, that is the reason for just using numbers. From my experience, temp1 seems to be the more useful, covering a greater range, so that is the one I'm using, displaying on my dashboard. But of course, it could be quite different with another Solax model...
Here is a recent day, temp2 did not even start below zero (even though it was cold in the morning, temp1 was about right), and both fall back to 0 at once when the inverter does not work:

Opera Snapshot_2024-01-11_210154_192 168 8 123

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxBindingConstants.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxBindingConstants.java:109]: This suggestion can wait for my next PR merge where I introduced Locale and time zone to the binding (from the framework. If you make the states an ENUM, I will later refactor them to use the translation files after my PR is merged...

**Comment by theater** [comment]
For a non-Java developer you did really well. :) I guess some programming experience you still have :)

**Comment by theater** [comment]
Btw if you don't have the capacity to do the requested changes, maybe you could give me permissions for this branch and I can do them for you. I expect that the reviewer will also request changes but it's up to you...

**Comment by realthk** [comment]
Thanks, I’d appreciate that! 
(yes, I mostly work on PHP projects, and as a hobby in Python, since Home Assistant)

**Comment by theater** [comment]
@realthk I've added two commits with the refactoring I was talking about and also added the update instructions.
Before I start I have also created a branch with the same name, ending with _backup, so you can go anytime to your original changes.
Would appreciate if you could test my changes whenever is possible for you.

**Comment by realthk** [comment]
@theater Just did a rebuild on your changes, and it works fine, thank you very much!

**Comment by theater** [comment]
OK. Great then the PR is ready for a review. Please tag me if someone requests some changes, so I can have a look and implement them...

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:126]: You can add possible states with the option tags to support UI and i18n

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/i18n/solax.properties
Comment (lsiepel) [inline bundles/org.openhab.binding.solax/src/main/resources/OH-INF/i18n/solax.properties:120]: the workmode (non code) channel seems t o be missing. Can you re-run i18n plugin?

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java:204]: I guess this is fine, in general i would personally prefer tekst channels for these kind of states. As there is no conversion/calculation so there is hardly any benefit in numbers and with a sstring also (future) text codes would be possible. Otherwise see general comment about the two channels.

**Review by lsiepel — CHANGES_REQUESTED** [review]
About the two channels `inverter-workmode` and `inverter-workmode-code` what is the benefit of seperating these? They seem to be allways in sync.
I guess it would be fine to drop the non-code variant and supply a option list with value/label pair: https://www.openhab.org/docs/developer/addons/config-xml.html#xml-structure-for-configuration-descriptions

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:126]: Hi, I prefer to keep that in the ENUM instead of the metadata unless it's strictly against the rules. 
There is a bigger change in my other PR (for the cloud connector) where I added Locale and timezone from the framework and my plan (already implemented there for the other relevant ENUMs) is to use the ENUM  value as an I18N key. Please let me know if this fits the requirement. I will refactor this one as a part of the other PR once this is merged (hopefully before the cloud connection PR)

File: bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java
Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java:204]: The text channel is there. I guess that Henrik added the code for additional information. Maybe we can switch this one to advanced, i.e. to be used only for rules and certain specific scenarios (even though I don't see any value if the relation is 1:1 code - enum value)

**Comment by realthk** [comment]
The reason why I thought it might be useful to have a text sensor, if not all Solax inverters use the same workmode codes for the same status, then it is possible to translate these.
But after googling (in this not so well documented subject) it seems the only difference is how many codes exists on a certain model but the most important first few statuses seem to share the same code for all inverter models, so the code might be enough, indeed.

**Comment by theater** [comment]
> The reason why I thought it might be useful to have a text sensor, if not all Solax inverters use the same workmode codes for the same status, then it is possible to translate these. But after googling (in this not so well documented subject) it seems the only difference is how many codes exists on a certain model but the most important first few statuses seem to share the same code for all inverter models, so the code might be enough, indeed.

Actually I have added all the statuses that I know about (from the cloud API document). Hopefully they're documented well enough...

File: bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml
Comment (lsiepel) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:126]: Not sure what your exact implementation would look like, but it sounds like a re-invention of base functionality. When this xml has the option tags, all translation is taken care of.  See the key/value pair enum/options here:
https://www.openhab.org/docs/developer/utils/i18n.html#thing-types

Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:126]: Agree. I will make the translations as options and will pass on the ENUM as a value, hopefully the framework will do it's job here. Looks much more clean. Question is do we leave the code as a numeric code and I do the translation part for the non-code channel (where I create an ENUM from the code)?
This one should be the translated as a string status:
`<channel id="inverter-workmode" typeId="inverter-workmode">`

Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:126]: One thing I didn't understand - if I add the options in the XML, do I always need to add them in the .properties file or it's assumed added and only the non-english translations must be added?
for example:
`<channel id="inverter-workmode" typeId="inverter-workmode">
   <options>
          <option value =  "NORMAL">Normal</option>
           .....
`
So do I have to add the following to the .properties file ? Something like this:
`thing-type.config.solax.inverter-workmode.language.option.NORMAL = Normal`

Comment (jlaur) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml:126]: They are needed in the `.properties` file in order to be translateable. You don't need to add them manually, you can run `mvn i18n:generate-default-translations` to update from the XML definitions - see https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

Comment (theater) [reply bundles/org.openhab.binding.solax/src/main/resources/OH-INF/thing/localConnectInverter.xml]: Hi, I think I got it right after a few failed tests now. Tested it with my inverter and looks to be working. What I did is..

1. Removed the wormode-code channel from the metadata completely
2. Added options tot he thing type of the workmode channel. Agree that these numbers do not make sense, therefore the channel should be workmode and not workmode code anymore.

**Comment by theater** [comment]
@jlaur, when you have the capacity and your schedule permits it, could we proceed as I think I have covered most if not all the concerns. 
Not trying to push you or anything, just want to make sure that you're not awaiting some changes from my end...

Cheers,
K.

**Comment by jlaur** [comment]
> @jlaur, when you have the capacity and your schedule permits it, could we proceed as I think I have covered most if not all the concerns. Not trying to push you or anything, just want to make sure that you're not awaiting some changes from my end...

I assume you mean @lsiepel? 🙂

**Comment by theater** [comment]
> > @jlaur, when you have the capacity and your schedule permits it, could we proceed as I think I have covered most if not all the concerns. Not trying to push you or anything, just want to make sure that you're not awaiting some changes from my end...
> 
> I assume you mean @lsiepel? 🙂

Apologies, mate! Not sure why I messed that up. Excuse me.

**Comment by theater** [comment]
@lsiepel could we continue the review when you have some time?

**Comment by lsiepel** [comment]
> @lsiepel could we continue the review when you have some time?

Sure! Hopefully this weekend, might be later. Have some busy days.

**Comment by theater** [comment]
I travel for the weekend so I will be able to react around Sunday. No
worries. No rush. Just wanted to check that it's not forgotten :)

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**
